pymarsys/connections.py

```diff
@@ -0,0 +1,159 @@
+        :return: headers.
+        """
+        if not other_headers:
+            other_headers = {}
+        nonce, created, password_digest = \
+            self.build_authentication_variables()
+
+        http_header = ','.join(
+            (
+                'UsernameToken Username="{}"'.format(self.username),
+                'PasswordDigest="{}"'.format(password_digest),
+                'Nonce="{}"'.format(nonce),
+                'Created="{}"'.format(created),
+            )
+        )
+        headers = {
+            'X-WSSE': http_header,
+            'Content-Type': 'application/json',
+            **other_headers,
+        }
+        return headers
+
+
+class SyncConnection(BaseConnection):
+    """
+    Synchronous connection for Ermasys or inherited-from BaseEndpoint objects.
+    """
+    def __init__(self, username, secret, uri=EMARSYS_URI):
+        super().__init__(username, secret, uri)
+
+    def make_call(self,
+                  method,
+                  endpoint,
+                  headers=None,
+                  payload=None):
+        """
+        Make an authenticated synchronous HTTP call to the Emarsys api using
+        the requests library.
+        :param endpoint: Emarsys' api endpoint.
+        :param method: HTTP method.
+        :param headers: HTTP headers.
+        :param payload: HTTP payload.
+        :param params: HTTP params.
+        :return: Dictionary with the result of the query.
+        """
+        if not payload:
+            payload = {}
+
+        url = os.path.join(self.uri, endpoint)
+        headers = self.build_headers(headers)
+        response = requests.request(
+            method,
+            url,
+            headers=headers,
+            json=payload
+        )
+        if response.ok:
+            logger.info(response.json())
+            return response.json()
+        else:
+            raise ApiCallError(
+                response.json()
+            )
+
+
+class AsyncConnection(BaseConnection):
+    """
+    Asynchronous connection for Ermasys or inherited-from BaseEndpoint objects.
+    """
+    def __init__(self, username, secret, uri=EMARSYS_URI):
+        super().__init__(username, secret, uri)
+
+    async def make_call(self,
+                        method,
+                        endpoint,
+                        headers=None,
+                        payload=None):
+        """
+        Make an authenticated asynchronous HTTP call to the Emarsys api using
+        the aiohttp library.
+        :param endpoint: Emarsys' api endpoint.
+        :param method: HTTP method.
+        :param headers: HTTP headers.
+        :param payload: HTTP payload.
+        :param params: HTTP params.
+        :return: Coroutine with the result of the query.
+        """
+        if not payload:
+            payload = {}
+
+        url = os.path.join(self.uri, endpoint)
+        headers = self.build_headers(headers)
+        async with ClientSession() as session:
```

Comment:
You probably shouldn't create a session for each request, you can probably instantiate the session when you initialize the connection and use it for future calls.